### PR TITLE
Add support for multi-document reference fields

### PR DESCRIPTION
**Pull request by eduardoboucas**
This PR implements #214, adding support for reference fields with multiple documents, based on the `settings.limit` property.

This is for non-media reference fields only — do we want to do the same for media fields, @mingard?

### CODE REVIEW

**Comment by mingard** [comment]
@eduardoboucas yes please - all Reference fields

**Comment by eduardoboucas** [comment]
Sure thing! It'll require some more work (e.g. requesting multiple signed URLs, handling multiple uploads, etc.), so once you're happy with this PR I suggest we merge it to avoid keeping `master` outdated for too long.

File: frontend/containers/DocumentListToolbar/DocumentListToolbar.jsx [hunk 1]
```diff
@@ -264,13 +266,14 @@ class DocumentListToolbar extends Component {
...
     // documents. For now, we just get the first selected document.
-    const selectedDocumentId = state.documents.selected[0]
-    const selectedDocument = documentsList.find(document => {
-      return document._id === selectedDocumentId
-    })
+    const selectedDocuments = state.documents.selected.map(documentId => {
+      return documentsList.find(document => {
+        return document._id === documentId
+      })
+    }).filter(Boolean)
```
Comment (mingard) [inline frontend/containers/DocumentListToolbar/DocumentListToolbar.jsx]: @eduardoboucas shorthand suggestion ;)
```javascript
const selectedDocuments = state.documents.selected
  .map(documentId => documentsList.find(document => document._id === documentId))
  .filter(Boolean)
```

File: frontend/containers/DocumentList/DocumentList.jsx [hunk 2]
```diff
@@ -417,9 +421,11 @@ class DocumentList extends Component {
   }
 
   renderField(fieldName, schema, value) {
-    const fieldType = schema.publish && schema.publish.subType ? schema.publish.subType : schema.type
+    const fieldType = schema.publish && schema.publish.subType ?
```
Comment (mingard) [inline frontend/containers/DocumentList/DocumentList.jsx]: I think I wrote this originally, but could we wrap multiple conditions perhaps?
```javascript
(schema.publish && schema.publish.subType)
```

File: frontend/components/FieldImage/FieldImageEdit.jsx [hunk 3]
```diff
@@ -174,26 +180,40 @@ export default class FieldImageEdit extends Component {
...
-    const reader = new FileReader()
-    
-    reader.onload = event => {
-      if (typeof onChange === 'function') {
-        onChange.call(this, schema._id, {
+    const singleFile = schema.settings && schema.settings.limit === 1
+
+    let processedFiles = []
+
+    for (let index = 0; index < files.length; index++) {
```
Comment (mingard) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: Just curious - any reason for not using this?
```javascript
files.forEach()
```

File: frontend/components/FieldImage/FieldImageEdit.jsx [hunk 4]
```diff
@@ -106,20 +106,25 @@ export default class FieldImageEdit extends Component {
       value
     } = this.props
     const displayName = schema.label || schema._id
-    const src = this.getImageSrc(value)
-    const isReference = schema.type === 'Reference'
     const fieldLocalType = schema.publish && schema.publish.subType ? schema.publish.subType : schema.type
     const href = buildUrl(...onBuildBaseUrl(), 'select', schema._id)
+    const isReference = schema.type === 'Reference'
+    const singleFile = schema.settings && schema.settings.limit === 1
+    const values = value && !(value instanceof Array) ? [value] : value
```
Comment (mingard) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: Suggestion:
```javascript
const values = (value && Array.isArray(value)) ? value : [value]
```

File: frontend/containers/DocumentListToolbar/DocumentListToolbar.jsx (hunk 1 above)
Comment (eduardoboucas) [inline frontend/containers/DocumentListToolbar/DocumentListToolbar.jsx]: I personally like to use shorthand arrow functions for simple and short statements. IMO having a `.find()` inside a shorthand `.map()` is unnecessarily cryptic.

File: frontend/containers/DocumentList/DocumentList.jsx (hunk 2 above)
Comment (eduardoboucas) [inline frontend/containers/DocumentList/DocumentList.jsx]: Sure, will change that.

File: frontend/containers/DocumentListToolbar/DocumentListToolbar.jsx (hunk 1 above)
Comment (mingard) [inline frontend/containers/DocumentListToolbar/DocumentListToolbar.jsx]: You could format it in another order. I believe it's more readable, but not a styleguide issue so up to you.

File: frontend/components/FieldImage/FieldImageEdit.jsx (hunk 3 above)
Comment (eduardoboucas) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: `files` is an instance of [FileList](https://developer.mozilla.org/en/docs/Web/API/FileList), not Array, so it doesn't have a `forEach` method.

Comment (mingard) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: ES6: `const files = [...FileList]`

File: frontend/components/FieldImage/FieldImageEdit.jsx (hunk 4 above)
Comment (eduardoboucas) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: If `value` is `null`, `values` would end up as `[null]` which evaluates to `true`.

Comment (mingard) [inline frontend/components/FieldImage/FieldImageEdit.jsx]:
```javascript
const values = (value && !Array.isArray(value)) ? [value] : value
```

Comment (mingard) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: Reason for suggestion is that `isArray` is considered more robust.

Comment (eduardoboucas) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: Sure. I think we're using `instanceof` in a few places, will need to change those too.

Comment (mingard) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: Sorry to make more work for you!

Comment (eduardoboucas) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: No worries, I'll change shortly.

Comment (eduardoboucas) [inline frontend/components/FieldImage/FieldImageEdit.jsx]: @mingard Done. Changed all instances I could find.

**Comment by eduardoboucas** [comment]
@mingard can we merge this?

**Pull request merged**